python/cudf/udf_cpp/strings/include/cudf/strings/udf/udf_string.hpp

```diff
@@ -545,6 +545,11 @@ class udf_string {
                               cudf::size_type nbytes);
 };
 
+struct managed_udf_string {
+  void* meminfo;
+  udf_string udf_str;
+};
```

Comment:
I'd rather this no live in this header file. I'm hoping to repurpose this for UDF needs for Spark and other users.
I see no reason it could not be moved to `udf_apis.hpp` though.

Reply:
Having a little trouble implementing this. Moving it to `udf_apis.hpp` outright seems not to work because this struct requires the complete definition of `udf_string` to compile:

```
error: incomplete type "cudf::strings::udf::udf_string" is not allowed
```

But, if you leave just a struct declaration in `udf_apis.hpp` and move the implementation to `udf_apis.cu` now the _cython_ complains it can't see the full `managed_udf_string` type.

```
invalid application of 'sizeof' to incomplete type 'cudf::strings::udf::managed_udf_string'
```

Thinking about how to handle this.

Reply:
You could perhaps put the `struct managed_udf_string` declaration in a new header file `managed_udf_string.hpp` and then include the `udf_string.hpp` there.
```
// copyright stuff
#pragma once
#include "udf_string.hpp"
struct managed_udf_string {
  void* meminfo;
  udf_string udf_str;
};
```
Then just include the `managed_udf_string.hpp` where you need to.

You could perhaps just create a single `managed_udf_string.cuh` that does the same thing but also has the definition:
```
__device__ void udf_str_dtor(void* udf_str, size_t size, void* dtor_info)
{
  auto ptr = reinterpret_cast<udf_string*>(udf_str);
  ptr->~udf_string();
}
```
I'm not sure if you would need both .hpp/.cuh or if you can just have the .cuh file have both.

Reply:
This has been done.